packages/app-page-builder/src/editor/plugins/elements/carousel/PeCarousel.tsx

```diff
@@ -0,0 +1,100 @@
+import React, { useRef, useEffect } from "react";
+import styled from "@emotion/styled";
+import { createRenderer, useRenderer, Elements } from "@webiny/app-page-builder-elements";
+import { Element } from "@webiny/app-page-builder-elements/types";
+import { useRecoilValue } from "recoil";
+import { elementWithChildrenByIdSelector } from "~/editor/recoil/modules";
+import { register } from "swiper/element/bundle";
+
+register();
+
+declare global {
+    // eslint-disable-next-line
+    namespace JSX {
+        interface IntrinsicElements {
+            "swiper-container": { [key: string]: any };
+            "swiper-slide": { [key: string]: any };
+        }
+    }
+}
+
+const PeCarouselWrapper = styled.div`
+    display: flex;
+    justify-content: center;
+    & swiper-slide {
+        text-align: center;
+        font-size: 16px;
+        background: #fff;
+        display: flex;
+        justify-content: center;
+        align-items: center;
+    }
+    & swiper-container {
+        max-width: 1100px;
+        --swiper-navigation-color: ${props => props.theme.styles.colors.color1};
+        --swiper-navigation-size: 40px;
+        --swiper-pagination-color: ${props => props.theme.styles.colors.color1};
+        --swiper-pagination-bottom: 0;
+    }
+    & .carousel-element-wrapper {
+        max-width: calc(95% - 41px);
+    }
+`;
+
+const PeCarousel = createRenderer(() => {
+    const swiperRef = useRef<any>(null);
+    const { getElement, meta } = useRenderer();
+    const element = getElement();
+    const elementWithChildren = useRecoilValue(
+        elementWithChildrenByIdSelector(element.id)
+    ) as Element;
+    const childrenElements = elementWithChildren?.elements;
+
+    const { arrowsToggle = true, bulletsToggle = true } = element?.data?.settings?.carousel;
+
+    const zIndex = meta.depth * 10;
+
+    useEffect(() => {
+        const swiperContainer = swiperRef?.current;
+        const params = {
+            rebuildOnUpdate: true
+        };
+        Object.assign(swiperContainer, params);
+        swiperContainer.initialize();
+    }, []);
+
+    useEffect(() => {
+        const swiperContainer = swiperRef?.current;
+        // By default swiper does not know that it has to rerender whenever we add a new slide
+        // So we have to update it manually whenever childrenElements.lenght is changing
+        // To do that we use swiper.update()
+        if (swiperContainer?.swiper?.update) {
+            swiperContainer.swiper.update();
+        }
+    }, [childrenElements.length]);
+
+    const navProps = {
+        navigation: arrowsToggle,
+        pagination: bulletsToggle
+    };
+
+    return (
+        <PeCarouselWrapper>
```

Comment:
Can't we just import the rendered from `app-page-builder-elements` ?

E.g. like we do it here: packages/app-page-builder/src/editor/plugins/elements/button/PeButton.tsx:8

Reply:
Maybe you are right!

But I have to clarify why we use separate react component `PeCarousel` instead of using a prerendering component `Carousel`.

1 - We need to use `useEffect` inside of `PeCarousel` to update Carousel whenever we add or remove a slide;
see here - https://github.com/webiny/webiny-js/pull/3276/files#diff-816934b2321ac03eb3e6bbdb09eced3c835698943e814901b642d999ec8965d2R68

2 - Also we use a ref to get an access to Swiper API methods;

3 - (Not actual anymore as [bug reported by Vitalii](https://github.com/nolimits4web/swiper/issues/6672) was already fixed in Swiper lib) Initially there were fixes to properly show a Carousel when arrows and bullets control settings were changed. (when we can configure in Carousel setting whether we want to show or hide arrows or bullets);

4 - All these things related to a Carousel settings changes or slides changes are possible only inside editor (published page will never see such case, as there is no data change after Swipper initialization);

We can make it as you have mentioned, but in such case I have to add `useEffect` and `ref` to the prerendering [Carousel](https://github.com/webiny/webiny-js/pull/3276/files#diff-009e5dab9be1a7d2a95e5ed3fe58f7e45a423b8858fd754f6f099591260e5a2eR79) component, even though published page does not need them.

Reply:
All good, I can see now essentially the components differ so much that it makes sense to have separate components. Resolving this convo.